### ISSUE TITLE
use xattrs for dos attributes

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#1344]
Recent region lock support patch moves
R/O dos attr from S_IWUSR to S_IWGRP.
Hidden attr is bound to dot and doesn't work.
Archive attr is bound to X bit.
System attr does not work.
Etc.

We need to move that to xattrs.
We need to use the notation of this patch of wine:
https://source.winehq.org/patches/data/163882
Which seems to be `user.DOSATTRIB` as a
name, and a value that represents the raw DOS
attr in a hex format.

Looks quite simple!
Anyone? :)

**Comment by stsp** [comment]
mtime should also be moved to xattrs for 2 reasons:
1. we don't want to change real mtime
2. DOS progs sometimes provide us with an invalid mtime, and DOS doesn't check that

**Comment by severach** [comment]
>R/O dos attr from S_IWUSR to S_IWGRP

So now all my files and folders need Linux perms be 664 instead of 644 and the umask must be 002. This seems like a major breaking change.

**Comment by stsp** [comment]
You don't need to do anything.
dosemu itself transfers 0644 to 0664.
Or doesn't it work?

**Comment by severach** [comment]
DOSEMU creates 664 files. Files created by Linux and old files all have 644 which now show as read-only in DOS. Old files I can fix. New files would require umask 002 or [setfacl](http://superuser.com/questions/151911/how-to-make-new-file-permission-inherit-from-the-parent-directory).

**Comment by stsp** [comment]
But mfs_reset() sets umask to 2 already, or?

**Comment by severach** [comment]
Inside DOSEMU yes. Files created by Linux such as files downloaded with Firefox and moved into DOS folders need chmod 664 to stop being read-only. We went from just a few select files being read only to all files before today being read-only.

**Comment by stsp** [comment]
Please test this as a work-around:
```
diff --git a/src/dosext/mfs/mfs.c b/src/dosext/mfs/mfs.c
index c0676a16d..175a59e76 100644
--- a/src/dosext/mfs/mfs.c
+++ b/src/dosext/mfs/mfs.c
@@ -1208,8 +1208,10 @@ int get_dos_attr(const char *fname,int mode,int hidden)
 
   if (S_ISDIR(mode) && !S_ISCHR(mode) && !S_ISBLK(mode))
     attr |= DIRECTORY;
+#if 0
   if (!(mode & S_IWGRP))
     attr |= READ_ONLY_FILE;
+#endif
   if (!(mode & S_IXGRP))
     attr |= ARCHIVE_NEEDED;
   if (hidden)
```

**Comment by severach** [comment]
That makes the 644 files writable but now `attrib` doesn't work. `attrib +r file`  changes the mode from 664 to 644 but the file isn't read-only. Files must be 444 to be read only which `attrib` can't do.

I'm not worried about me. I can chmod, run this patch, or run an old version; and my files don't really need to be read only. The current Group RO method makes DOSEMU much harder to deploy to DOS users than it already is. The xattr solution needs to be sooner than later.

**Comment by stsp** [comment]
I applied the work-around, as having
no R/O support is no better than no
System or Hidden support, and none
of them currently work.
So at least the regression is now not
that big.
Yes, someone needs to code up xattrs.

**Comment by severach** [comment]
Something is wrong in the new code. I'm getting files that can't be deleted.

````
D:\voters20\gck21l>dir *.fxp
 Volume in drive D is OME CHRIS D
 Directory of D:\VOTERS20\GCK21L
 
SUMM     FXP        12,317  11-27-20  7:02p
         1 file(s)         12,317 bytes
         0 dir(s)    262,140 Mega bytes free
D:\voters20\gck21l>del *.fxp
D:\VOTERS20\GCK21L\summ.fxp: No such file or directory
no file removed.
D:\voters20\gck21l>dir *.fxp
 Volume in drive D is OME CHRIS D
 
 Directory of D:\VOTERS20\GCK21L
 
SUMM     FXP        12,317  11-27-20  7:02p
         1 file(s)         12,317 bytes
         0 dir(s)    262,140 Mega bytes free

***

[chris@fujitsu gck21l]$ ls -l *.fxp
-rw-rw-r-- 1 chris users 12317 Nov 27 19:02 summ.fxp
````

Quit and reload DOSEMU to delete the file. Linux also has no problem deleting the file.

`lsof` doesn't show the fxp file as open in either `xsel` #1325 or `dosemu.bin`. The file is open by `xsel` because I can `killall xsel` then delete the file from DOS.

**Comment by stsp** [comment]
This should now be fixed.

@andrewbird could you please disable
the attribute tests for now?

**Comment by andrewbird** [comment]
Mmm, you could just prefix `test_mfs_ds2_set_fattrs` in test_dos.py with `x`, but there looks to be something substantial to the other failures.

**Comment by andrewbird** [comment]
Just looking at adding XATTRs support. Do I need to handle the readonly bit in XATTRs or can it fallback to filesystem?

**Comment by stsp** [comment]
Everything should go to xattrs, even mtime.
We can't lock w/o rw access.

**Comment by andrewbird** [comment]
But surely directory can come from the fiesystem and we just ignore volume label?

**Comment by stsp** [comment]
I guess so.

**Comment by andrewbird** [comment]
I have a first stab at this https://github.com/andrewbird/dosemu2/tree/travis-ci-46, and I'm able to store and retrieve DOS attrs via xattrs. I re-enabled the setfattrs tests and they pass. I do have a different failing test though
`python test/test_dos.py MSDOS622TestCase.test_mfs_ds3_share_open_setfattrs`
I think I haven't ripped out enough of the old filesystem permission method, I don't really understand what the GRP perms were doing there.

**Comment by andrewbird** [comment]
Found it, though I'm sure there are lots of other fixes to add yet

**Comment by andrewbird** [comment]
One other thought, Instead of using xattrs, do you think it would be possible to encode the DOS attributes in an additional acl to each file? So perhaps a file could have:
~~~
getfacl: Removing leading '/' from absolute path names
# file: tes.fil
# owner: ajb
# group: sphere
user::rw-
user:dosattr20:rwx
user:dosattr53:rwx
group::rw-
mask::rw-
other::---
~~~
We'd create the users dosattr{20,53}bits at install time, Nobody could ever login as them, and we'd be able to set and unset those acls as owner of the file without affecting any linux access.

**Comment by stsp** [comment]
I am not sure messing with creating the
users is a good idea. Just what does it
give us over xattrs? Incompats with wine,
inability to set attrs like system/hidden,
and just for what?

I also think you should add the read-only
check to `do_mfs_open()`.

**Comment by stsp** [comment]
Thanks for your work on this!
Together with an sdl-ttf patch, you
would plug two major holes.
Which would be a huge boost, as
I don't think we have a lot of "major
holes" today.

**Comment by andrewbird** [comment]
>  Just what does it
give us over xattrs?

Well it works over NFS now, and so not having to wait for some maybe distant version of NFSv4.2 to be ratified, implemented and become commonplace. It doesn't matter for me as I now have a workaround, but I wonder if anyone else uses NFS?

> inability to set attrs like system/hidden

I was proposing that the acl above gave us 6 bits, which if values are maintained should cover up to and including the archive bit at 0x20

> I also think you should add the read-only
check to do_mfs_open().

But didn't you establish that files had to be unix RW for exclusive locking to work?

**Comment by andrewbird** [comment]
> Together with an sdl-ttf patch, ...

That's stalled at the moment for me. I was finding that doing the development of it within the dosemu video backend confusing as there's so much I don't know. I started to write a simple standalone SDL app to test things like the scaling etc, and at least proved that the tearing seen during window enlargement that I was concerned about was a 'feature' of my bare minimum Radeon X1300 card(2005).  I'd upgrade it, but that's a can of worms I'd rather not open, if only for the humongous power usage that modern cards want!  At some point I'll come back to it, but I want to catch up with the testing updates I have in mind first.

**Comment by stsp** [comment]
> I was proposing that the acl above gave us 6 bits

Which is why you add 2 users, and for mtime -
another 5? :) Well, not the best idea IMO.

> But didn't you establish that files had to be unix RW for exclusive locking to work?

You just need to check that xattrs allow
write if write_requested==1, otherwise
returning an error.

> tearing seen during window enlargement that I was concerned about was a 'feature' of my bare minimum Radeon X1300 

Not sure its possible.
We use software renderer.
Its independent of any card.

**Comment by andrewbird** [comment]
> Which is why you add 2 users, and for mtime -
another 5? :) Well, not the best idea IMO.

Yes, well, perhaps not.

>You just need to check that xattrs allow
write if write_requested==1, otherwise
returning an error.

Okay.

> Not sure its possible.
We use software renderer.
Its independent of any card.

Then perhaps I managed to recreate exactly the same cr*pness in my standalone app too! At least it's good for trying to find out what's wrong with it.

**Comment by stsp** [comment]
Please upload your example.

**Comment by andrewbird** [comment]
I uploaded it to the proper thread.

**Comment by andrewbird** [comment]
I was wondering what the proper behaviour of setting an attribute on an open file should be, so I added a temporary test to see what happens on FAT. Firstly the behaviour differs according to whether share is loaded, but seeing as FDPP always has share available, I loaded it for the other DOSes. The test is here https://github.com/andrewbird/dosemu2/tree/travis-ci-46.

##### DR-DOS 7.01
~~~
C:\>testit
 
C:\>c:\share
SHARE R2.02    File sharing support.
Copyright (c) 1988,1996 Caldera, Inc.  All rights reserved.
 
C:\>d:
D:\>set LFN=n
 
D:\>c:\defattr
INFO: Create success = 'MYTEST.FIL'
INFO: Getfattr(1) attr=0x20
INFO: Setfattr success
INFO: Getfattr(2) attr=0x21

# note file is successfully closed as per normal
PASS: results as expected
 
D:\>rem end
~~~

##### MS-DOS 6.22
~~~
C:\>testit.bat
testit.bat

C:\>c:\share

C:\>d:

D:\>set LFN=n

D:\>c:\defattr
INFO: Create success = 'MYTEST.FIL'
INFO: Getfattr(1) attr=0x20
INFO: Setfattr success
INFO: Getfattr(2) attr=0x21
FAIL: Close failed, err=-1

D:\>rem end
~~~

##### FR-DOS 1.20
~~~
C:\>testit.bat                                                                  
testit.bat                                                                      
C:\>c:\share                                                                    
SHARE installed.
C:\>d:                                                                          
D:\>set LFN=n                                                                   
D:\>c:\defattr                                                                  
INFO: Create success = 'MYTEST.FIL'                                             
INFO: Getfattr(1) attr=0x20                                                     
INFO: Setfattr success                                                          
INFO: Getfattr(2) attr=0x20                                                     
FAIL: Getfattr(2) expected (0x21), got (0x20)                                   
D:\>rem end                                                                     
~~~

##### FDPP
~~~
C:\>testit.bat                                                                  
testit.bat

C:\>rem Internal share
C:\>d:
D:\>set LFN=n
D:\>c:\defattr
INFO: Create success = 'MYTEST.FIL'                                             
INFO: Getfattr(1) attr=0x20                                                     
FAIL: Setfattr failed, err=5                                                    
~~~

So which behaviour is the correct one for FAT, and for MFS we have to consider if other processes have the file open too?
- DR-DOS seems quite normal, changing the ATTRs on an open file shouldn't affect any processes that have it open already
- MS-DOS was broadly the same, except that it automatically closed the file, why?
- FR-DOS just ignored the attr change, but didn't return an error on the set, not sure if the file was closed or not.
- FDPP refused the set with ACCESS_DENIED.

Thoughts?

**Comment by andrewbird** [comment]
rbil has something to say about it http://www.ctyme.com/intr/rb-2803.htm

**Comment by stsp** [comment]
I think one process being able to
close file of another process, is a
nonsense.
As for ACCESS_DENIED, I think
this is what share does (perhaps
also on dr-dos?)

**Comment by andrewbird** [comment]
> I think one process being able to
close file of another process, is a
nonsense.

As do I. I didn't even consider the two process case, but I did have share loaded. In my example the same process opens the file, then attempts to setattr, and finally close it.

The MS-DOS closed the file after the very same process did setattr on it. This seems a little strange. But perhaps the whole concept of doing setattr with an open file is strange anyway.

My wip for MFS attrs currently behaves as per FDPP, that is the setattr is refused with ACCESS_DENIED. I think that's because it looks at the open_files array, though that's not good enough for the multi dosemu process, which I suppose would need to check for locks. 

> (perhaps
also on dr-dos?)

In the DR-DOS FAT test above the setattr completed, the attr actually changed, and the open status didn't change. That sounds great but I suspect that it was all done rather naively without any share checking at all.

Perhaps I need to do a two process share test like we did with delete / rename and open. With a matrix of results according to share open modes.
